File: src/app/core/tradeModes.ts
/**
 * Canonical trade-mode rules (TS mirror of python/cicada_nn/trade_modes.py).
 *
 * Single source of truth for the five trade styles. The Python and TS tables
 * are kept identical by `scripts/verify-trade-mode-parity.ts`, run via
 * `npm run verify-all`.
 *
 * If you change a number here, change it in `trade_modes.py` too. Otherwise
 * the parity script will fail in CI.
 */

import type { TradeStyle } from './types';

export type SLManagement =
  | 'static'
  | 'trail_after_1r'
  | 'be_then_trail'
  | 'trail_from_entry';

export type TPManagement =
  | 'fixed'
  | 'partial_1r_rest_tp'
  | 'partial_1r_rest_2r';

export interface TradeModeRules {
  style: TradeStyle;
  timeframes: readonly string[];
  minHoldBars: number;
  minTpAtr: number;
  minSlAtr: number;
  maxSlAtr: number;
  slManagement: SLManagement;
  tpManagement: TPManagement;
  entryConfirmation: string;
  exitTrigger: string;
  maxConcurrent: number;
  confidenceThreshold: number;
  mt5Magic: number;
}

export const TRADE_MODES: Record<TradeStyle, TradeModeRules> = {
  scalping: {
    style: 'scalping',
    timeframes: ['M1', 'M5'],
    minHoldBars: 3,
    minTpAtr: 0.5,
    minSlAtr: 0.3,
    maxSlAtr: 1.0,
    slManagement: 'static',
    tpManagement: 'fixed',
    entryConfirmation: 'PA + map zone',
    exitTrigger: 'TP/SL/reversal after min hold',
    maxConcurrent: 3,
    confidenceThreshold: 0.6,
    mt5Magic: 1001,
  },
  day: {
    style: 'day',
    timeframes: ['M15', 'M30', 'H1'],
    minHoldBars: 6,
    minTpAtr: 1.0,
    minSlAtr: 0.6,
    maxSlAtr: 2.0,
    slManagement: 'trail_after_1r',
    tpManagement: 'partial_1r_rest_tp',
    entryConfirmation: 'PA + map zone + 1 indicator',
    exitTrigger: 'TP/SL/regime break',
    maxConcurrent: 2,
    confidenceThreshold: 0.65,
    mt5Magic: 1002,
  },
  medium_swing: {
    style: 'medium_swing',
    timeframes: ['H1', 'H4'],
    minHoldBars: 8,
    minTpAtr: 1.5,
    minSlAtr: 0.8,
    maxSlAtr: 3.0,
    slManagement: 'be_then_trail',
    tpManagement: 'partial_1r_rest_2r',
    entryConfirmation: 'Map zone + momentum',
    exitTrigger: 'TP/SL/structure break',
    maxConcurrent: 2,
    confidenceThreshold: 0.7,
    mt5Magic: 1003,
  },
  swing: {
    style: 'swing',
    timeframes: ['H4', 'D1'],
    minHoldBars: 12,
    minTpAtr: 2.0,
    minSlAtr: 1.0,
    maxSlAtr: 4.0,
    slManagement: 'be_then_trail',
    tpManagement: 'partial_1r_rest_2r',
    entryConfirmation: 'Map zone + momentum + structure',
    exitTrigger: 'TP/SL/structure break',
    maxConcurrent: 2,
    confidenceThreshold: 0.7,
    mt5Magic: 1004,
  },
  sniper: {
    style: 'sniper',
    timeframes: ['M15', 'M30', 'H1'],
    minHoldBars: 6,
    minTpAtr: 1.5,
    minSlAtr: 0.8,
    maxSlAtr: 2.5,
    slManagement: 'trail_from_entry',
    tpManagement: 'fixed',
    entryConfirmation: '2+ S/R confluence + map zone',
    exitTrigger: 'TP or SL only',
    maxConcurrent: 1,
    confidenceThreshold: 0.8,
    mt5Magic: 1005,
  },
};

export const ALL_TRADE_MODES: readonly TradeStyle[] = [
  'scalping',
  'day',
  'medium_swing',
  'swing',
  'sniper',
];

export function getRules(style: TradeStyle): TradeModeRules {
  const r = TRADE_MODES[style];
  if (!r) throw new Error(`Unknown trade style: ${style}`);
  return r;
}

// ─── Validation (mirror of validate_order in trade_modes.py) ────────────────

export type RejectReason =
  | 'ok'
  | 'tp_too_tight'
  | 'sl_too_tight'
  | 'sl_too_wide'
  | 'confidence_below_threshold'
  | 'max_concurrent_exceeded'
  | 'min_hold_not_elapsed'
  | 'invalid_signal';

export interface OrderSignal {
  side: 'LONG' | 'SHORT';
  entryPrice: number;
  stopLoss: number;
  takeProfit: number;
  confidence: number;
}

export interface ValidationResult {
  ok: boolean;
  reason: RejectReason;
  detail: string;
}

function reject(reason: RejectReason, detail: string): ValidationResult {
  return { ok: false, reason, detail };
}

function accept(): ValidationResult {
  return { ok: true, reason: 'ok', detail: '' };
}

export function validateOrder(
  rules: TradeModeRules,
  signal: OrderSignal,
  atr: number,
  nConcurrent: number,
  barsSinceLastOpen: number | null = null
): ValidationResult {
  if (signal.entryPrice <= 0 || atr <= 0) {
    return reject('invalid_signal', `entry=${signal.entryPrice}, atr=${atr}`);
  }
  if (signal.confidence < rules.confidenceThreshold) {
    return reject(
      'confidence_below_threshold',
      `${signal.confidence.toFixed(3)} < ${rules.confidenceThreshold.toFixed(3)}`
    );
  }
  if (nConcurrent >= rules.maxConcurrent) {
    return reject('max_concurrent_exceeded', `${nConcurrent} >= ${rules.maxConcurrent}`);
  }
  if (barsSinceLastOpen !== null && barsSinceLastOpen < rules.minHoldBars) {
    return reject('min_hold_not_elapsed', `${barsSinceLastOpen} < ${rules.minHoldBars}`);
  }
  const slDistance = Math.abs(signal.entryPrice - signal.stopLoss);
  const tpDistance = Math.abs(signal.entryPrice - signal.takeProfit);
  const slAtrMult = slDistance / atr;
  const tpAtrMult = tpDistance / atr;
  if (tpAtrMult < rules.minTpAtr) {
    return reject('tp_too_tight', `${tpAtrMult.toFixed(3)}xATR < ${rules.minTpAtr}`);
  }
  if (slAtrMult < rules.minSlAtr) {
    return reject('sl_too_tight', `${slAtrMult.toFixed(3)}xATR < ${rules.minSlAtr}`);
  }
  if (slAtrMult > rules.maxSlAtr) {
    return reject('sl_too_wide', `${slAtrMult.toFixed(3)}xATR > ${rules.maxSlAtr}`);
  }
  return accept();
}

File: python/cicada_nn/trade_modes.py
"""
Canonical trade-mode rules for CICADA-5453.

Single source of truth for the five trade styles (scalping / day / medium_swing
/ swing / sniper). Mirrored in TypeScript at ``src/app/core/tradeModes.ts``;
parity verified by ``scripts/verify-trade-mode-parity.ts`` (run via
``npm run verify-all``).

Modes were previously sharing one validation path (the daemon's risk gate),
which let a parameter set picked for SCALPING coerce a SWING signal. This
module replaces that with strict per-mode rules and a pure ``validate_order``
function that returns a dataclass result — never silently mutates the order.
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum
from typing import Literal


TradeStyle = Literal["scalping", "day", "medium_swing", "swing", "sniper"]


class SLManagement(str, Enum):
    STATIC = "static"
    TRAIL_AFTER_1R = "trail_after_1r"
    BE_THEN_TRAIL = "be_then_trail"          # breakeven at +1R, then trail
    TRAIL_FROM_ENTRY = "trail_from_entry"


class TPManagement(str, Enum):
    FIXED = "fixed"
    PARTIAL_1R_REST_TP = "partial_1r_rest_tp"
    PARTIAL_1R_REST_2R = "partial_1r_rest_2r"


@dataclass(frozen=True)
class TradeModeRules:
    style: TradeStyle
    timeframes: tuple[str, ...]
    min_hold_bars: int
    min_tp_atr: float
    min_sl_atr: float
    max_sl_atr: float
    sl_management: SLManagement
    tp_management: TPManagement
    entry_confirmation: str
    exit_trigger: str
    max_concurrent: int
    confidence_threshold: float
    mt5_magic: int


TRADE_MODES: dict[TradeStyle, TradeModeRules] = {
    "scalping": TradeModeRules(
        style="scalping",
        timeframes=("M1", "M5"),
        min_hold_bars=3,
        min_tp_atr=0.5,
        min_sl_atr=0.3,
        max_sl_atr=1.0,
        sl_management=SLManagement.STATIC,
        tp_management=TPManagement.FIXED,
        entry_confirmation="PA + map zone",
        exit_trigger="TP/SL/reversal after min hold",
        max_concurrent=3,
        confidence_threshold=0.60,
        mt5_magic=1001,
    ),
    "day": TradeModeRules(
        style="day",
        timeframes=("M15", "M30", "H1"),
        min_hold_bars=6,
        min_tp_atr=1.0,
        min_sl_atr=0.6,
        max_sl_atr=2.0,
        sl_management=SLManagement.TRAIL_AFTER_1R,
        tp_management=TPManagement.PARTIAL_1R_REST_TP,
        entry_confirmation="PA + map zone + 1 indicator",
        exit_trigger="TP/SL/regime break",
        max_concurrent=2,
        confidence_threshold=0.65,
        mt5_magic=1002,
    ),
    "medium_swing": TradeModeRules(
        style="medium_swing",
        timeframes=("H1", "H4"),
        min_hold_bars=8,
        min_tp_atr=1.5,
        min_sl_atr=0.8,
        max_sl_atr=3.0,
        sl_management=SLManagement.BE_THEN_TRAIL,
        tp_management=TPManagement.PARTIAL_1R_REST_2R,
        entry_confirmation="Map zone + momentum",
        exit_trigger="TP/SL/structure break",
        max_concurrent=2,
        confidence_threshold=0.70,
        mt5_magic=1003,
    ),
    "swing": TradeModeRules(
        style="swing",
        timeframes=("H4", "D1"),
        min_hold_bars=12,
        min_tp_atr=2.0,
        min_sl_atr=1.0,
        max_sl_atr=4.0,
        sl_management=SLManagement.BE_THEN_TRAIL,
        tp_management=TPManagement.PARTIAL_1R_REST_2R,
        entry_confirmation="Map zone + momentum + structure",
        exit_trigger="TP/SL/structure break",
        max_concurrent=2,
        confidence_threshold=0.70,
        mt5_magic=1004,
    ),
    "sniper": TradeModeRules(
        style="sniper",
        timeframes=("M15", "M30", "H1"),
        min_hold_bars=6,
        min_tp_atr=1.5,
        min_sl_atr=0.8,
        max_sl_atr=2.5,
        sl_management=SLManagement.TRAIL_FROM_ENTRY,
        tp_management=TPManagement.FIXED,
        entry_confirmation="2+ S/R confluence + map zone",
        exit_trigger="TP or SL only",
        max_concurrent=1,
        confidence_threshold=0.80,
        mt5_magic=1005,
    ),
}


ALL_STYLES: tuple[TradeStyle, ...] = ("scalping", "day", "medium_swing", "swing", "sniper")


def get_rules(style: str) -> TradeModeRules:
    if style not in TRADE_MODES:
        raise KeyError(f"Unknown trade style: {style!r}. Known: {list(TRADE_MODES)}")
    return TRADE_MODES[style]  # type: ignore[index]


# ─── Validation ─────────────────────────────────────────────────────────────


class RejectReason(str, Enum):
    OK = "ok"
    TP_TOO_TIGHT = "tp_too_tight"
    SL_TOO_TIGHT = "sl_too_tight"
    SL_TOO_WIDE = "sl_too_wide"
    CONFIDENCE_BELOW_THRESHOLD = "confidence_below_threshold"
    MAX_CONCURRENT_EXCEEDED = "max_concurrent_exceeded"
    MIN_HOLD_NOT_ELAPSED = "min_hold_not_elapsed"
    INVALID_SIGNAL = "invalid_signal"
    # Stage 2A: latency-aware gates per spec lines 1418-1431.
    BASELINE_NOT_ESTABLISHED = "baseline_not_established"
    RTT_STALE = "rtt_stale"
    LATENCY_ELEVATED = "latency_elevated"
    LATENCY_ANOMALY = "latency_anomaly"
    LATENCY_SEVERE = "latency_severe"
    LATENCY_EXTREME = "latency_extreme"
    UNKNOWN_LATENCY_REASON = "unknown_latency_reason"


# Map the string reasons returned by ``LatencyModel.get_trade_gate`` to typed
# ``RejectReason`` values. Kept here (rather than in latency_model) so the
# validation contract has a single source of truth.
_LATENCY_REASON_MAP: dict[str, RejectReason] = {
    "OK": RejectReason.OK,
    "BASELINE_NOT_ESTABLISHED": RejectReason.BASELINE_NOT_ESTABLISHED,
    "RTT_STALE": RejectReason.RTT_STALE,
    "LATENCY_ELEVATED": RejectReason.LATENCY_ELEVATED,
    "LATENCY_ANOMALY": RejectReason.LATENCY_ANOMALY,
    "LATENCY_SEVERE": RejectReason.LATENCY_SEVERE,
    "LATENCY_EXTREME": RejectReason.LATENCY_EXTREME,
}


def latency_reason_to_reject(reason: str) -> RejectReason:
    """Translate a ``TradeGate.reason`` string to a ``RejectReason``."""
    return _LATENCY_REASON_MAP.get(reason, RejectReason.UNKNOWN_LATENCY_REASON)


@dataclass(frozen=True)
class OrderSignal:
    """Pure data sent to ``validate_order`` — no mutation, no I/O."""

    side: Literal["LONG", "SHORT"]
    entry_price: float
    stop_loss: float
    take_profit: float
    confidence: float


@dataclass(frozen=True)
class ValidationResult:
    ok: bool
    reason: RejectReason
    detail: str = ""

    @classmethod
    def accept(cls) -> "ValidationResult":
        return cls(ok=True, reason=RejectReason.OK)

    @classmethod
    def reject(cls, reason: RejectReason, detail: str = "") -> "ValidationResult":
        return cls(ok=False, reason=reason, detail=detail)


def _abs_distance(a: float, b: float) -> float:
    return abs(a - b)


def validate_order(
    rules: TradeModeRules,
    signal: OrderSignal,
    atr: float,
    n_concurrent: int,
    bars_since_last_open: int | None = None,
    *,
    latency_baseline_valid: bool | None = None,
    latency_gate_allowed: bool | None = None,
    latency_gate_reason: str | None = None,
    latency_gate_detail: str | None = None,
) -> ValidationResult:
    """Validate an order against per-mode rules.

    Pure function — never mutates ``signal`` or ``rules``. Caller must log a
    ``REJECTED`` row when ``ok is False`` and never coerce the order to fit.

    :param rules: Mode rules from ``TRADE_MODES``.
    :param signal: The order intent.
    :param atr: Current ATR (in price units, same as entry/stop/TP).
    :param n_concurrent: How many positions this bot already has open in this
        mode. Caller enforces this against ``max_concurrent``.
    :param bars_since_last_open: Bars elapsed since the last entry on this
        bot. ``None`` skips the gate (e.g., fresh bot with no prior position).
    :param latency_baseline_valid: Result of
        ``latency_model.is_baseline_valid()`` — spec check #6. ``None``
        skips (used in unit tests not exercising live latency).
    :param latency_gate_allowed: Result of
        ``latency_model.get_trade_gate(mode).allowed`` — spec check #7.
        ``None`` skips. When ``False``, ``latency_gate_reason`` provides the
        typed rejection.
    :param latency_gate_reason: ``TradeGate.reason`` string from
        ``latency_model``. Used to map to a ``RejectReason`` when check #7
        fails.
    :param latency_gate_detail: Optional human-readable detail (e.g. the
        gate's ``recommendation`` field) carried into the rejection.
    """
    if signal.entry_price <= 0 or atr <= 0:
        return ValidationResult.reject(
            RejectReason.INVALID_SIGNAL,
            f"entry_price={signal.entry_price}, atr={atr}",
        )

    if signal.confidence < rules.confidence_threshold:
        return ValidationResult.reject(
            RejectReason.CONFIDENCE_BELOW_THRESHOLD,
            f"{signal.confidence:.3f} < {rules.confidence_threshold:.3f}",
        )

    if n_concurrent >= rules.max_concurrent:
        return ValidationResult.reject(
            RejectReason.MAX_CONCURRENT_EXCEEDED,
            f"{n_concurrent} >= {rules.max_concurrent}",
        )

    if bars_since_last_open is not None and bars_since_last_open < rules.min_hold_bars:
        return ValidationResult.reject(
            RejectReason.MIN_HOLD_NOT_ELAPSED,
            f"{bars_since_last_open} < {rules.min_hold_bars}",
        )

    sl_distance = _abs_distance(signal.entry_price, signal.stop_loss)
    tp_distance = _abs_distance(signal.entry_price, signal.take_profit)
    sl_atr_mult = sl_distance / atr
    tp_atr_mult = tp_distance / atr

    if tp_atr_mult < rules.min_tp_atr:
        return ValidationResult.reject(
            RejectReason.TP_TOO_TIGHT,
            f"{tp_atr_mult:.3f}xATR < {rules.min_tp_atr:.3f}",
        )
    if sl_atr_mult < rules.min_sl_atr:
        return ValidationResult.reject(
            RejectReason.SL_TOO_TIGHT,
            f"{sl_atr_mult:.3f}xATR < {rules.min_sl_atr:.3f}",
        )
    if sl_atr_mult > rules.max_sl_atr:
        return ValidationResult.reject(
            RejectReason.SL_TOO_WIDE,
            f"{sl_atr_mult:.3f}xATR > {rules.max_sl_atr:.3f}",
        )

    # ── Stage 2A: latency gates (spec lines 1418-1431). Run AFTER the
    # mode-shape checks so a tight TP doesn't get masked by a latency
    # spike — the operator wants the structural reason first.
    if latency_baseline_valid is False:
        return ValidationResult.reject(
            RejectReason.BASELINE_NOT_ESTABLISHED,
            latency_gate_detail or "latency baseline still warming up",
        )
    if latency_gate_allowed is False:
        reason = latency_reason_to_reject(latency_gate_reason or "")
        return ValidationResult.reject(
            reason,
            latency_gate_detail or (latency_gate_reason or "latency gate blocked"),
        )

    return ValidationResult.accept()
